python/rascaline/power_spectrum.py

```diff
@@ -0,0 +1,201 @@
+                        np.intersect1d(block1.samples, block2.samples).tolist()
+                    ),
+                )
+
+                if len(common_samples) == 0:
+                    continue
+
+                spx1 = slice_block(block1, axis="samples", labels=common_samples)
+                spx2 = slice_block(block2, axis="samples", labels=common_samples)
+            else:
+                spx1 = block1
+                spx2 = block2
+
+            # Avoid doubly computing / storing invariants that are
+            # the same by symmetry of the neighbor species.
+            # Example: Neighbor species (Na, Cl) produces the same
+            # invariants as (Cl, Na), meaning that only one set
+            # of invariants needs to be used.
+            # If the two sets of expansion coefficients are different,
+            # this does not apply
+            if sn1 > sn2 and use_same_spherical_expansions:
+                continue
+
+            factor = 1 / sqrt(2 * l1 + 1)
+            if sn1 != sn2:
+                factor *= sqrt(2)
+
+            properties = Labels(
+                names=[f"{name}1" for name in spx1.properties.names]
+                + [f"{name}2" for name in spx2.properties.names],
+                values=np.array(
+                    [
+                        properties1.tolist() + properties2.tolist()
+                        for properties1 in spx1.properties
+                        for properties2 in spx2.properties
+                    ],
+                    dtype=np.int32,
+                ),
+            )
+
+            # Compute the invariants by summation and store the results
+            # operation: ima, imb -> iab
+            data = factor * np.matmul(spx1.values.swapaxes(1, 2), spx2.values)
+
+            block = TensorBlock(
+                values=data.reshape(data.shape[0], -1),
+                samples=spx1.samples,
+                components=[],
+                properties=properties,
+            )
+
+            n_properties = block.values.shape[1]
+
+            for parameter in spx1.gradients_list():
+                if parameter != "positions":
+                    raise NotImplementedError(
+                        "power spectrum currently only supports position gradients"
+                    )
+
+                gradient1 = spx1.gradient(parameter)
+                gradient2 = spx2.gradient(parameter)
+
+                if len(gradient1.samples) == 0 or len(gradient2.samples) == 0:
+                    continue
+
+                # The "sample" dimension in the power spectrum gradient samples do not
+                # necessarily matches the "sample" dimension in the spherical expansion
+                # gradient samples. We create new samples by creating a union between
+                # the two gradient samples.
+                gradients_samples_values = np.unique(
+                    np.concatenate([gradient1.samples, gradient2.samples])
```

Comment:
I think this is wrong, and has the same issue we recently fixed in the Rust implementation

spx1 samples could contain atom 1 at position 0, while spx2 samples could contain atom 1 at position 2 (if the neighbor species are different in the single spx case, or just with different cutoffs if we give this function two separate spx)

This means the gradient sample (0, 0, 1) in block 1 and (2, 0, 1) in block 2 would actually refer to the same gradient, and this would fail to find it. Similarly, I could see a case where identical gradient samples refer to different atoms/neighbors.